### PR TITLE
Dev 13478 Additional typing for a helpful TS Review reference configuration

### PR DESCRIPTION
**Pull request by sergiovp**

### CODE REVIEW

File: src/ui/FilterForm.tsx [hunk 1]
```diff
@@ -1,33 +1,34 @@
...
 
+import { FormFeedbackByName, FormValueByName } from 'fontoxml-design-system/src/types';
 import t from 'fontoxml-localization/src/t';
 
 import useNestedCheckboxesForFilterOptions from './useNestedCheckboxesForFilterOptions';
+import { $TSFixMeAny, AnnotationErrorType } from '../types';
 
-function determineParentFieldValue(subFieldNames, valueByNameForUI) {
+function determineParentFieldValue(subFieldNames: string[], valueByNameForUI: $TSFixMeAny) {
 	return subFieldNames.reduce((value, subFieldName) => {
```
Comment (albertosml) [inline src/ui/FilterForm.tsx]: `return subFieldNames.reduce<boolean>((value, subFieldName) => {`

File: src/ui/FilterFormSummaryChips.tsx [hunk 2]
```diff
@@ -13,6 +13,21 @@ import * as React from 'react';
...
+import { FormFeedbackByName } from 'fontoxml-design-system/src/types';
+import { AnnotationErrorType } from '../types';
+import { CardContentComponentProps } from 'fontoxml-feedback/src/types';
+
+type Props = {
+	error: AnnotationErrorType;
+	feedbackByName: FormFeedbackByName;
+	isDisabled: boolean;
+	isSubmitting: boolean;
+	onChange: CardContentComponentProps['onReviewAnnotationFormSubmit'];
```
Comment (albertosml) [inline src/ui/FilterFormSummaryChips.tsx]: Even the `onChange` function and the submit function for the review annotation form have the same type, I think that they are not related because they do different actions, so I would set another type like `onChangeCallback` from `fontoxml-design-system` library or a custom type as `onChange(valueByName: FormValueByName): Promise<void>;`

File: src/ui/FilterForm.tsx [hunk 3]
```diff
@@ -36,6 +37,17 @@ function determineParentFieldValue(subFieldNames, valueByNameForUI) {
...
 
+type Props = {
+	error: AnnotationErrorType;
+	feedbackByName: FormFeedbackByName;
+	isSubmitting: boolean;
+	onFieldChange(...args: unknown[]): void;
+	onInitialize(): void;
+	productContent: string;
+	showFeedback: boolean;
+	valueByName: FormValueByName;
```
Comment (albertosml) [inline src/ui/FilterForm.tsx]: All the values stored in the `valueByName` variable will be boolean, so the type can be changed to `{ [key: string]: boolean }`.

File: src/ui/useAuthorAndTimestampLabel.tsx [hunk 4]
```diff
@@ -13,21 +13,24 @@ const configuredScope = configurationManager.get('scope');
...
  * author field is not set on the given reviewAnnotationOrReply.
  *
- * @return {object} Object containing the 'author' and the 'timestamp' keys.
+ * @returns Object containing the 'author' and the 'timestamp' keys.
  * @react
  */
 export default function useAuthorAndTimestampLabel(
-	reviewAnnotationOrReply,
-	isReviewAnnotationResolved,
+	reviewAnnotationOrReply: CardContentComponentProps['reviewAnnotation'],
```
Comment (albertosml) [inline src/ui/useAuthorAndTimestampLabel.tsx]: It can also receive a reply, not only a review annotation.

File: src/ui/comment/CommentAddOrEditForm.tsx [hunk 5]
```diff
@@ -126,12 +133,19 @@ function CommentAddOrEditFormContent({
 	);
 }
 
+type Props = {
+	reviewAnnotation: CardContentComponentProps['reviewAnnotation'];
```
Comment (albertosml) [inline src/ui/comment/CommentAddOrEditForm.tsx:140]: Change it to `AnnotationDataModel` instead of `ReviewAnnotation` type.

File: src/ui/proposal/ProposalCardContent.tsx [hunk 6]
```diff
@@ -96,8 +97,8 @@ function ProposalCardContent({
 
 	// Replace the whole card if the reviewAnnotation.error is acknowledgeable.
 	if (
-		reviewAnnotation.error &&
-		reviewAnnotation.error.recovery === RecoveryOption.ACKNOWLEDGEABLE
+		typeof reviewAnnotation.error !== 'number' &&
+		reviewAnnotation?.error.recovery === RecoveryOption.ACKNOWLEDGEABLE
```
Comment (albertosml) [inline src/ui/proposal/ProposalCardContent.tsx]: I think this is causing hardcrash in proposals.

File: src/ui/shared/CardHeader.tsx [hunk 7]
```diff
@@ -25,19 +26,38 @@ import AuthorAndTimestampLabel from '../AuthorAndTimestampLabel';
...
+	hasReplyInNonIdleBusyState: boolean;
+	isSelectedToShare: CardContentComponentProps['isSelectedToShare'];
+	onReviewAnnotationEdit: CardContentComponentProps['onReviewAnnotationEdit'];
+	onReviewAnnotationRemove: CardContentComponentProps['onReviewAnnotationRemove'];
+	onReviewAnnotationResolve: CardContentComponentProps['onReviewAnnotationResolve'];
+	onReviewAnnotationShare: CardContentComponentProps['onReviewAnnotationShare'];
+	onReviewAnnotationShareAddRemoveToggle: CardContentComponentProps['onReviewAnnotationShareAddRemoveToggle'];
+	onReviewAnnotationShowInCreatedContext: CardContentComponentProps['onReviewAnnotationShowInCreatedContext'];
+	onReviewAnnotationShowInResolvedContext: CardContentComponentProps['onReviewAnnotationShowInResolvedContext'];
+	reviewAnnotation: CardContentComponentProps['reviewAnnotation'];
```
Comment (albertosml) [inline src/ui/shared/CardHeader.tsx:59]: Set this type to `CardContentComponentProps['reviewAnnotation'] & { targets: Target[] }`.

File: src/ui/shared/Reply.tsx [hunk 8]
```diff
@@ -10,14 +10,28 @@ import {
...
 import TruncatedText from './TruncatedText';
 
+type Props = {
+	reviewAnnotation: CardContentComponentProps['reviewAnnotation']; 
+	onCancelRetryRemove: CardContentComponentProps['onReplyFormCancel'];
+	onHide: CardContentComponentProps['onReplyErrorHide'];
+	onRefresh: CardContentComponentProps['onReplyRefresh'];
+	onRemove: CardContentComponentProps['onReplyRemove'];
+	onShowEditForm: CardContentComponentProps['onReplyEdit'];
+	reply: ReplyType;
```
Comment (albertosml) [inline src/ui/shared/Reply.tsx:32]: `ReplyType` is unknown, you have to import it. But, I think that the right type is `Reply`, which is located on `fontoxml-feedback` and has to be exported.

File: src/ui/shared/ReplyForm.tsx [hunk 9]
```diff
@@ -53,17 +61,30 @@ function validateReplyField(value) {
...
+	onSubmit, 
+	reply, 
+}: {
+	isSubmitDisabled: boolean;
+	onCancelButtonClick: () => void;
+	onFocusableRef(): void;
+	onHideLinkClick: CardContentComponentProps['onReplyErrorHide'];
+	onRefreshLinkClick: CardContentComponentProps['onReplyRefresh'];
+	onSubmit: CardContentComponentProps['onReplyFormSubmit'];
+	reply: ReplyType;
```
Comment (albertosml) [inline src/ui/shared/ReplyForm.tsx:78]: `ReplyType` does not exist.

File: src/ui/shared/ReplyForm.tsx [hunk 10]
```diff
@@ -128,7 +149,15 @@ function ReplyFormContent({
 	);
 }
 
-function ReplyForm({ onCancel, onHide, onRefresh, onSubmit, reply }) {
+type Props = {
+	onCancel: CardContentComponentProps['onReplyFormCancel'];
+	onHide: CardContentComponentProps['onReplyErrorHide'];
+	onRefresh: CardContentComponentProps['onReplyRefresh'];
+	onSubmit: CardContentComponentProps['onReplyFormSubmit'];
+	reply: ReplyType;
```
Comment (albertosml) [inline src/ui/shared/ReplyForm.tsx:158]: `ReplyType` does not exist.

**Review by albertosml — CHANGES_REQUESTED** [review]
Changes (apart from the ones added in comments):

- You should use the <> generic type for "useState" and "useRef".
- Missing return type for each component/function.
- Hardcrash when adding a proposal.
- Comment form is disabled.
- In `src/AuthorAndTimestampLabel.tsx` file, the variable `authorLabelScrollWidth` must be a number initialized with 0, not with a boolean.

Extra changes:
- In `src/ui/FilterFormSummaryChips.tsx` file, the `tooltipContent` property must be declared in the Chip FDS component props. Shall we do it on this story?
- In `src/ui/MastheadForReview.tsx` file, `ReviewLogo` component is missing onClick property.
- In `src/ui/proposal/ProposalAddOrEditForm.tsx` and `src/ui/comment/CommentAddOrEditForm.tsx` files, I observe that `TextAreaWithDiff` and `TextArea` components, respectively, are passing a `ref` property, which does not exist on those components; I would suggest to check if the `onRef` can be passed or not. Same problem on `src/ui/shared/ReplyForm.tsx`.
- In `src/ui/proposal/ProposalCardContent.tsx` file, we should cast to string the values passed to the `Diff` component through `value` property.
- In `src/ui/shared/CardHeader.tsx` file, shall we rename `color` property from `Icon` FDS component to `colorName`?
- Submit function is being passed on the `onClick` function from a button on `src/ui/shared/ReplyForm.tsx`, `src/ui/shared/ResolveForm.tsx` and `src/ui/shared/AddOrEditFormFooter.tsx` files.
- In `src/ui/shared/TruncatedText.tsx` file, does the useTruncation hook really need parameters if they are not passed to the unique invocation?

File: src/ui/comment/CommentAddOrEditForm.tsx (hunk 5 above)
Comment (sergiovp) [reply src/ui/comment/CommentAddOrEditForm.tsx:140]: That would cause other errors. I think the best option will be to add targets attribute as optional to the `ReviewAnnotation` type.

File: src/ui/proposal/ProposalCardContent.tsx (hunk 6 above)
Comment (sergiovp) [reply src/ui/proposal/ProposalCardContent.tsx]: Yes, you are right Alberto, thanks, it should be changed to `reviewAnnotation.error?.recovery`

File: src/ui/shared/CardHeader.tsx (hunk 7 above)
Comment (sergiovp) [reply src/ui/shared/CardHeader.tsx:59]: You are right, and also targets should be optional.

File: src/ui/shared/Reply.tsx (hunk 8 above)
Comment (sergiovp) [reply src/ui/shared/Reply.tsx:32]: Yes, you're right.

File: src/ui/shared/AddOrEditFormFooter.tsx [hunk 11]
```diff
@@ -2,21 +2,40 @@ import { Button, Flex } from 'fds/components';
...
 import t from 'fontoxml-localization/src/t';
 
 import ResponsiveButtonSpacer from './ResponsiveButtonSpacer';
+import { AnnotationErrorType } from 'fontoxml-review-reference-configuration/src/types';
 
-function determineSaveButtonLabel(error, isLoading) {
+function determineSaveButtonLabel(
+	error: AnnotationErrorType,
+	isLoading: boolean
+) {
```
Comment (albertosml) [inline src/ui/shared/AddOrEditFormFooter.tsx]: Missing return type.

File: src/ui/shared/AddOrEditFormFooter.tsx [hunk 12]
```diff
@@ -2,21 +2,40 @@ import { Button, Flex } from 'fds/components';
...
+	error:  AnnotationErrorType;
+	isDisabled: boolean;
+	isLoading: boolean;
+	isSubmitDisabled: boolean;
+	onCancel: CardContentComponentProps['onReviewAnnotationFormCancel'];
+	onReviewAnnotationRefresh: CardContentComponentProps['onReviewAnnotationRefresh'];
+	onSubmit: CardContentComponentProps['onReviewAnnotationFormSubmit'];
+};
+
+const paddingSize = { top: 'l' };
```
Comment (albertosml) [inline src/ui/shared/AddOrEditFormFooter.tsx]: To fix the TS error from line 58, we need to indicate that this is a `PaddingSize` object.

```
const paddingSize: PaddingSize = { top: 'l' };
```

File: src/ui/shared/CardHeader.tsx [hunk 13]
```diff
@@ -25,19 +27,39 @@ import AuthorAndTimestampLabel from '../AuthorAndTimestampLabel';
...
 import { CARD_HEADER_HEIGHT } from './../constants';
 import FeedbackContextType from 'fontoxml-feedback/src/FeedbackContextType';
+import { AnnotationErrorType } from 'fontoxml-review-reference-configuration/src/types';
 
-function determineShareButtonLabel(reviewAnnotation, error, isLoading) {
+function determineShareButtonLabel(
+	reviewAnnotation: CardContentComponentProps['reviewAnnotation'],
+	error: AnnotationErrorType,
+	isLoading: boolean)
+{
```
Comment (albertosml) [inline src/ui/shared/CardHeader.tsx]: Missing return type.

File: src/ui/shared/Reply.tsx [hunk 14]
```diff
@@ -10,14 +10,29 @@ import {
 import * as React from 'react';
 
 import ErrorToast from 'fontoxml-feedback/src/ErrorToast';
-import { BusyState } from 'fontoxml-feedback/src/types';
+import {
+	BusyState,
+	CardContentComponentProps,
+	Reply as ReplyType
```
Comment (albertosml) [inline src/ui/shared/Reply.tsx:16]: Why do you rename the reply type? I would directly leave it as Reply.

File: src/ui/shared/ReplyForm.tsx [hunk 15]
```diff
@@ -10,15 +10,25 @@ import * as React from 'react';
...
-function determineSaveButtonLabel(error, isEditing, isLoading) {
-	if (error && error.recovery === RecoveryOption.RETRYABLE) {
+import { AnnotationErrorType } from 'fontoxml-review-reference-configuration/src/types';
+import { FormFeedback } from 'fontoxml-design-system/src/types';
+
+function determineSaveButtonLabel(
+	error: AnnotationErrorType,
+	isEditing: boolean,
+	isLoading: boolean
+) {
```
Comment (albertosml) [inline src/ui/shared/ReplyForm.tsx]: Missing return type.

File: src/ui/shared/ResolveForm.tsx [hunk 16]
```diff
@@ -26,26 +27,32 @@ import t from 'fontoxml-localization/src/t';
 import resolutions from '../feedbackResolutions';
 import ResponsiveButtonSpacer from './ResponsiveButtonSpacer';
 
+import { FormFeedback, FormValueByName } from 'fontoxml-design-system/src/types';
+import { AnnotationErrorType } from 'fontoxml-review-reference-configuration/src/types';
+
 const rows = { minimum: 2, maximum: 6 };
 
-function determineSaveButtonLabel(error, isLoading) {
+function determineSaveButtonLabel(error: AnnotationErrorType, isLoading: boolean) {
```
Comment (albertosml) [inline src/ui/shared/ResolveForm.tsx]: Missing return type.

File: src/ui/FilterForm.tsx [hunk 17]
```diff
@@ -1,33 +1,34 @@
 import { Block, Checkbox, Flex, Label } from 'fds/components';
 import * as React from 'react';
 
+import { FormFeedbackByName, FormValueByName } from 'fontoxml-design-system/src/types';
```
Comment (albertosml) [inline src/ui/FilterForm.tsx]: `FormValueByName` is unused.

Comment (ThomasBrekelmans) [reply src/ui/FilterForm.tsx]: And should be used to type valueByNameForUI: $TSFixMeAny I think

File: src/ui/FilterForm.tsx [hunk 18]
```diff
@@ -1,33 +1,34 @@
...
 import t from 'fontoxml-localization/src/t';
 
 import useNestedCheckboxesForFilterOptions from './useNestedCheckboxesForFilterOptions';
+import { $TSFixMeAny, AnnotationErrorType } from '../types';
 
-function determineParentFieldValue(subFieldNames, valueByNameForUI) {
-	return subFieldNames.reduce((value, subFieldName) => {
+function determineParentFieldValue(subFieldNames: string[], valueByNameForUI: $TSFixMeAny) {
```
Comment (ThomasBrekelmans) [inline src/ui/FilterForm.tsx]: I think you can use FormValueByName instead of $TSFixMeAny here?

File: src/ui/shared/AddOrEditFormFooter.tsx (hunk 11 above)
Comment (ThomasBrekelmans) [reply src/ui/shared/AddOrEditFormFooter.tsx]: : string 😄

File: src/ui/shared/CardHeader.tsx (hunk 13 above)
Comment (ThomasBrekelmans) [reply src/ui/shared/CardHeader.tsx]: : string again 😄

File: src/ui/shared/CardHeader.tsx [hunk 19]
```diff
@@ -376,7 +400,7 @@ export default function CardHeader({
 													.resolution ===
 													'accepted' && (
 													<Icon
-														color="inherit"
+														colorName="inherit"
```
Comment (ThomasBrekelmans) [inline src/ui/shared/CardHeader.tsx:402]: Nice fix

File: src/ui/shared/CardHeader.tsx [hunk 20]
```diff
@@ -386,7 +410,7 @@ export default function CardHeader({
 													.resolution ===
 													'rejected' && (
 													<Icon
-														color="inherit"
+														colorName="inherit"
```
Comment (ThomasBrekelmans) [inline src/ui/shared/CardHeader.tsx:412]: Nice fix

**Review by ThomasBrekelmans — APPROVED** [review]
I'm pre-approving. All of the feedback is pretty minor. Feel free to merge when you're done with it.

File: src/ui/shared/Reply.tsx (hunk 14 above)
Comment (sergiovp) [reply src/ui/shared/Reply.tsx:16]: Having a type whose name is the same as the name of a function causes errors:
image

File: src/ui/shared/Replies.tsx [hunk 21]
```diff
@@ -73,6 +73,7 @@ const Replies: React.FC<Props> = ({
 
 				return (
 					<Reply
+						key={reply.id}
```
Comment (sergiovp) [inline src/ui/shared/Replies.tsx]: This prop was added to avoid the following error:
image

**Review by albertosml — APPROVED** [review]
